### PR TITLE
Polars: 0.12 documentation

### LINKED ISSUE #2101: Polars: update docs website to cover new OpenDP 0.12 functionality

### PR DESCRIPTION
**Pull request by Shoeboxam**
Close #2101

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/2200?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2200** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2200?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/opendp/opendp/2200?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2199** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2199?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by mccalluc — APPROVED** [review]
This is a pretty quick skim, and I'm now going to check out this branch and build the docs locally, just to see what the user would see. It would be fine with me if you wanted to merge this down the chain, and if there are any more issues, they can be addressed before the merge to main. (Though I think you usually do the merges in the opposite direction to keep PRs from getting too large.) -- In any case, I've marked it approved, but I'm also planning on looking at it some more.

I do wonder if the more reference-y parts would be better under the API user guide, but I'm not confident about that: Your call.

File: docs/source/getting-started/tabular-data/data-types.ipynb
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/data-types.ipynb:121]: Once the data is ingest, it isn't possible to have different time zones, because Polars stores the time zone and time unit in the schema, and the data itself is just a vector of 64-bit ints. I'd imagine Polars would throw an error if the time zones don't match, but that's Polars' concern, not ours.

File: docs/source/getting-started/tabular-data/expressions/aggregation.ipynb
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/expressions/aggregation.ipynb:9]: My thinking for these OpenDP expression sections is that they just cover OpenDP-specific behavior, like how a transformation may add or remove domain descriptors, or options that are explicitly not allowed.

File: docs/source/getting-started/tabular-data/expressions/aggregation.ipynb
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/expressions/aggregation.ipynb:237]: No, the noise scale is 180 and the true value is zero. If the noise is negative, it will return zero.

**Comment by Shoeboxam** [comment]
You're right that the Polars docs are now more thorough that just "getting started", but it also feels a shame to split them up.

**Review by mccalluc — COMMENTED** [review]
Looking at the docs as rendered, some more comments. I think a lot of them aren't about changes in this PR, and so they might be new issues, if we want to think about them at all.

#### Prefer if the expr reference could be moved to API User Guide

But if you do believe it should stay here, could you limit the depth of the ToC on getting-started/index.html, and on getting-started/tabular-data/index.html, just so it can fit in the window?

#### OpenDP Polars vs typical Polars

> As shown in these code snippets, OpenDP Polars differs from typical Polars in three ways

by this point, the code that you are referring to is more than a screen away. Consider including short example blocks under "How you specify the data" to remind the reader what we're talking about.

Or, alternatively, move this discussion up, or interleave with the code above.

#### Overflow and stability

> The OpenDP Library ensures that privacy guarantees take into account the potential for overflow and/or numerical instability. For this reason, many statistics require a known upper bound on how many records can be present in the data

There's a whole lot of detail here, and in the sections that follow, and I wonder if it might be overwhelming for a new user. I wonder if this could be shorter, with a link to more explanation in the API user-guide?

> the worst-case error from summing floating-point numbers also becomes very large

We care about this, but I'm not sure a getting-started reader does.

#### Margins

> ... in the form of a margin. A margin is used to describe certain properties that a potential adversary would already know about the data.

If someone with a stats background already has a definition for "margin" in their head, is our definition going to connect to it? Or is it going to feel like we're redefining terminology?

#### Don't use private data for imputation

> Do not use private data to calculate imputed values or bounds

This could be expanded. How did we know that zero is a good value to impute? When would it not be good? What are the plusses/minuses of filtering nulls vs. imputation?

Maybe cover this elsewhere and link to it?

#### Means

> As of the current release of the OpenDP Library, it is recommended to estimate means by separately releasing sum and count estimates

It feels like there are two cases: if you have the lengths, and if you don't, but if you do have lengths, `.dp.mean` is great.

(Also, avoid passive voice.)

> remove null values (denoted by 99).

Maybe "Not applicable", to be consistent with the previous page, and avoid confusion with actual nulls.

> `query_work_hours.release().collect()`

We should complete this example and actually calculate the dp mean from these two values.

> 1-stable transformations

This probably won't be meaningful to a new user. This discussion is too detailed.

#### Less apologetic

> We have been working to improve the inferential logic to avoid the need to pull preprocessing outside of the library

> As of the current release of the OpenDP Library

I think we should just focus on what the library does now. If there's a tracking issue for an improvement, it might be linked, but for the most part let's not bother the user with our plans.

#### Median

> The more candidates you pass, the higher the computational overhead (growing logarithmically) and memory overhead (growing linearly).

These both seem like really low rates of growth: Do they even need to be mentioned here?

#### Quantile + Multistats

Just combine, and present as a generalization of median?

(Update the outline at the top to be consistent with this presentation.)

#### defining delta and stable

I think the first mention is in the middle of this paragraph:

> OpenDP calibrates this filtering threshold such that the probability of releasing a sensitive grouping key is no greater than the privacy parameter delta (δ).

If this is going to be a definition of delta, then this sentence should be flipped around.

> This is why the set of released keys are considered “stable”:

I don't think we've defined stable before this point.

I'm not sure we _should_ define these here: How far can the getting-started docs help users get started without diving into these details?

#### lets -> let's (where it means "let us")

#### Microdata

Instead of 
> Microdata (data where each row corresponds to a different individual) is seldom already in the form you need it in

could we just say 
> Data is seldom already in the form you need it in

It feels like we're introducing jargon.

#### Shadowing

> would no longer apply to the new, shadowed, HWUSUAL column after

Should be "shadowing" rather than "shadowed"?

#### Partitioning

> Filtering preserves margin descriptors for the greatest number of partitions, greatest number of influenced partitions, greatest number of records in a partition, and greatest number of records contributed to a partition

Lots of words! For an introduction, maybe just focus on what is lost. 

Maybe a table of what is kept and lost? Not sure how useful it would be.

#### Gaussian noise

> you can easily switch to zCDP

I think this is way too deep for the getting-started, and when it is introduced, it would be good to describe the problem that we're solving.

#### Expression index

```
:py:class`DPExpr`
```
should be?
```
:py:class:`DPExpr`
```

**Review by mccalluc — COMMENTED** [review]
Just have the Expression Index left. (Reading from the rendered version: Sorry not to put comments on diff)

My main concern is still whether this belongs in getting-started: Because it is systematic, and not a tutorial, and makes some assumptions about DP knowledge, parts may be better in the API user guide.

We might also think about pulling out the table of methods from api/python/opendp.extras.polars.html (I don't think we're going to be keeping it up to date), so the API docs can just focus on function-by-function information, and for a higher level view, the API docs would link to these pages in the API user guide. 

#### Expression Index

> The OpenDP Library operates based on an allowlist of recognized expressions

This is a description of the implementation. Maybe rephrase: "OpenDP supports many, but not all, Polars expressions, and adds extensions under the `dp` namespace."

> The following sections align one-to-one with the Polars API documentation structure

Is there a top-level Polars page we could link to? Structurally, https://docs.pola.rs/api/python/stable/reference/expressions/index.html seems like it should be the place, but it lists all the functions, and doesn't have the groups that you have here.

#### Aggregation

> The most common aggregators like length, sum, mean, median and quantile are covered in the introductory essential statistics notebook.

Not an absolute, but I'd like getting-started to be readable start-to-end without too many internal links. (Links out to the API reference are different.) I think we can assume to some extent that it is read in order. (One reason to keep the scope from getting too wide.)

> Counting queries tend to be relatively easy to release with high utility in differential privacy.

I'd avoid describing things as "easy": If it really is easy for the user, they don't need us to confirm it -- And if they aren't finding it easy, it really doesn't help to try to tell them otherwise!

I think this may also be describing the underlying mechanism: The library should be adding the abstractions that hide this from the user, to some extent.

> Just like how you can differentiate between COUNT(*) and COUNT(HWUSUAL) in SQL

I would drop this. Even if the reader knows SQL, the distinction is subtle. If you want to keep it, make it more like "By the way, this is analogous to the distinction in SQL between ..."

> total number of elements in the HWUSUAL column (including nulls)

Should be "excluding"? Or maybe I'm confused.

> These two statistics are equivalent

But not if they treat nulls differently?

> unbounded-DP (add/remove).

I don't think this has been defined at this point. I'd be cautious about introducing too much DP terminology in getting started.

> Noise added to a count can make the count go negative, but since the output data type is an unsigned integer, the library may return zero

Oh: I had wondered about this with DP Wizard. Is this really the behavior we want? I think I remember asking about negative counts a year ago, and I think the response then was that we didn't want to clip them because it lost information: If a DP count is more negative, that is saying that it is even more unlikely that the true value is non-zero. (Though, depending on the audience, you may still want to clip to avoid explaining negative counts.)

#### Boolean

Maybe introduce with something like: "OpenDP supports a subset of the [Polars boolean functions](https://docs.pola.rs/api/python/stable/reference/expressions/boolean.html)." 

> Since these expressions are row-by-row, they may be used in row-by-row or aggregation contexts - is_null - is_not_null - is_finite - is_not_finite - is_nan - is_not_nan

This is rendered all inline. Change to `*` for RST list. (I could write a little test to catch MD syntax in RST, but we should also check how it renders. File an issue if you feel like it.)

> We may lift this limitation in the future by rewriting such joint expressions into individual expressions, as Polars does

If we do want to talk about our TODOs, consider adding a link to the corresponding issue as a way to get feedback about whether this would be useful.

> they will not be supported until after landing support for selectors that span multiple columns. Until then

Too much detail about our roadmap. Most users just want to know what works now.

> ```
> # not recommended, OpenDP will reject this joint expression over multiple columns
> single_expr = pl.col([f"METHOD{l}" for l in "ABCDE"]).fill_null(0).dp.sum((0, 9))
> ```

Doctest showing the error might be nice. (Also a good check for us that the error is human readable.)

#### Manipulation

Particularly in this section, since it's just going function by function, I wonder if it would make sense just to fill in more details in the API reference docs?

#### Operators

```diff
- All conjunction, comparison and binary operators in the linked documentation
+ All Polars [conjunction](https://docs.pola.rs/api/python/stable/reference/expressions/operators.html#conjunction), [comparison](https://docs.pola.rs/api/python/stable/reference/expressions/operators.html#comparison), and [binary](https://docs.pola.rs/api/python/stable/reference/expressions/operators.html#binary) operators
```

#### String

> Dates can be parsed from strings via .str.strptime, and its variants .str.to_date, .str.to_datetime, and .str.to_time.

If this becomes RST, link to API reference docs.

#### Temporal

This page and the preceding one are short and related. Consider giving up the strict parallel to the Polars docs and combining these into one.

**Comment by Shoeboxam** [comment]
*Prefer if the expr reference could be moved to API User Guide*
moved!

*OpenDP Polars vs typical Polars*
added!


*Overflow and stability*

Will leave this as-is for now, but once modular noise mechanisms are added the discussion could be delayed.

*Margins*

In hindsight I think "Grouping" would have been a better name than "Margin".
It hasn't been enough of a pain for me to change it.

*Don't use private data for imputation*

Changed the fill value and added more explanation.

*Means*
Updated!

*Quantile + Multistats*
Updated!


*defining delta and stable*
I feel like this helps to explain the connection between the parameters and potential loss of partitions.


*Microdata*
Changed


*Shadowing*
Now shadowing

*Partitioning*
Reduced.

*Gaussian noise*
Updated.

*Expression index*
Fixed

*Expression Index*
Not that I know of.

*Aggregation*
Len includes nulls.

The negative behavior is how things currently are, 
but it would be good to have a setting to cast the dtype.

*Columns*
The error when you specify multiple columns is along the lines of: 
`Expr not recognized cols([A, B]).`

*Temporal*
Both will become longer, so will IMO feel more natural as we continue to develop.

**Review by mccalluc — COMMENTED** [review]
Great. A few edits: Let me know if pushing to github (either on this branch or a new one) would be useful.

[polars-doc.patch](https://github.com/user-attachments/files/18199393/polars-doc.patch)

I also think we would really benefit by porting these notebooks to RST, but doesn't need to block this PR (URLs will remain the same).
- Filed #2205

**Review by mccalluc — APPROVED** [review]
Approved!

**Pull request merged**
